### ISSUE TITLE
MambaND is just Identity ? So this is just 1D Mamba and Vision Mamba ?

### ISSUE THREAD

**Issue by sendeniz** [jacklishufan/Mamba-ND#10]
In several files (e.g., image_classification/src/mamba.py and the VisionMamba implementation), MambaND and related components are currently stubbed out:

```
MambaND = nn.Identity  # TODO: Cleanup and release
Block2D = SplitHead2D = nn.Identity
```

meaning its usually just 1D Mamba or Vision Mamba and not really an N-Dimensional SSM. It may thus be easier to just use vision mamba, since this repo is not nearly as well documented.

**Comment by sendeniz** [comment]
We evaluated Mamba-ND on CIFAR-10 as a visual representation learner and compared it against S4ND under similar settings.
Key findings:

Mamba-ND achieves ~91.8% test accuracy vs ~94% for S4ND, a gap that persisted across extensive hyperparameter sweeps (depth, embed dim, dropout, learning rate, weight decay)

- Increasing patch size degrades performance consistently; 
- patch size = 1 (native resolution) slowls down compute by a lot

Mamba-ND is better understood as a scan-based wrapper around Mamba/VMamba with multi-axial scanning rather than a jointly parameterized N-dimensional state space model. It outperforms vanilla Vision Mamba on this task, likely due to the multi-axis scanning mechanism, but does not match S4ND.

The core distinction from S4ND is the shift from time-invariant (HiPPO/S4/S4ND) to time-variant (selective) state space formulation — trading representational capacity on spatially structured data for memory efficiency and input-dependent dynamics

Takeaway: For spatially structured benchmarks like CIFAR-10, S4ND's jointly parameterized multidimensional formulation appears to be a meaningful advantage. Mamba-ND's efficiency gains come at a measurable accuracy cost, making S4ND the preferred choice when memory is not the primary constraint.

You can try num_layers = 12, ti improvements from 12 to 12 // 2 are marginal however. Could try embed_dims = 128 * 6, but this lead to instability, likely requiering either 1) the grad clip or 2) lower learning rate. Takes quite a bit of time to test, so I wont be trying this out. 

```
import torch
from image_classification.src.mamba import Mamba2DModel
from torchvision import datasets, transforms
from torch.utils.data import DataLoader
import torch.nn as nn

batch_size = 32 * 4 
epochs = 50
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
])

train_data = datasets.CIFAR10('./data', train=True, download=True, transform=transform)
test_data  = datasets.CIFAR10('./data', train=False, download=True, transform=transform)
train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True,  num_workers=4)
test_loader  = DataLoader(test_data,  batch_size=batch_size, shuffle=False, num_workers=4)

class SimpleMamba(nn.Module):
    def __init__(self):
        super().__init__()
        self.backbone = Mamba2DModel(
            arch='small', 
            with_cls_token=False, 
            drop_path_rate=0.1, 
            drop_rate=0.1,
            final_norm=True, 
            img_size=32, 
            patch_size=1, 
            in_channels=3, 
            embed_dims=128*3,
            num_layers=12//2,
            out_type='avg_featmap', 
            is_2d=False, 
            has_reverse=True,
            has_transpose=True, 
            fused_add_norm=False, 
            constant_dim=True,
            downsample=(), 
            d_state=16,
            use_v2=False,
        )
        self.head = nn.Linear(128*3, 10)

    def forward(self, x):
        return self.head(self.backbone(x)[0])
    

model = SimpleMamba().to(device)
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=1e-4, betas=(0.9, 0.999))

scheduler = torch.optim.lr_scheduler.OneCycleLR(
    optimizer,
    max_lr=3e-4,
    steps_per_epoch=len(train_loader),
    epochs=epochs,
    pct_start=3 / epochs,       # 3-epoch warmup out of 50
    anneal_strategy='cos',
    div_factor=10,              # initial_lr = max_lr / 10 = 1e-4
    final_div_factor=1e3,       # final_lr  = initial_lr / 1e3 ≈ 1e-7
)

print(f"Training on {device}")

for epoch in range(epochs):
    model.train()
    correct, total = 0, 0
    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)
        optimizer.zero_grad()
        outputs = model(images)
        loss = criterion(outputs, labels)
        loss.backward()
        #torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=2.0)
        optimizer.step()
        scheduler.step()        # called per batch with OneCycleLR

        _, predicted = outputs.max(1)
        total   += labels.size(0)
        correct += predicted.eq(labels).sum().item()

    train_acc  = 100. * correct / total
    current_lr = scheduler.get_last_lr()[0]

    model.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for images, labels in test_loader:
            images, labels = images.to(device), labels.to(device)
            outputs = model(images)
            _, predicted = outputs.max(1)
            total   += labels.size(0)
            correct += predicted.eq(labels).sum().item()

    test_acc = 100. * correct / total
    print(f"Epoch {epoch+1:02d}/{epochs} | LR: {current_lr:.2e} | "
          f"Train: {train_acc:.2f}% | Test: {test_acc:.2f}%")

print("\nDone!")
```
